internal/server/server.go

```diff
@@ -563,6 +563,12 @@ func (s *Server) endUpload(ctx context.Context, src snapshot.SourceInfo) {
 	s.parallelSnapshotsChanged.Signal()
 }
 
+// GetRepository get the repository for this server.  This is used to ensure
+// repositories are closed on shutdown.  Also symmetric to SetRepository.
+func (s *Server) GetRepository() repo.Repository {
+	return s.rep
```

Comment:
My suggestion would be to try to solve the `ensure repositories are closed on shutdown` problem in a separate PR, so that it doesn't get mixed u with your signal handling changes. It seems like the imbalanced closing of repos is an existing problem. So, unless we are introducing a new problem, we might eb able to do it separately without causing any breakage, what do you think ?